RecoTracker/LSTCore/src/alpaka/TrackCandidate.h

```diff
@@ -13,121 +17,31 @@
-    }
-  };
-
-  template <typename TDev>
-  struct TrackCandidatesBuffer {
-    Buf<TDev, short> trackCandidateType_buf;
-    Buf<TDev, unsigned int> directObjectIndices_buf;
-    Buf<TDev, unsigned int> objectIndices_buf;
-    Buf<TDev, unsigned int> nTrackCandidates_buf;
-    Buf<TDev, unsigned int> nTrackCandidatespT3_buf;
-    Buf<TDev, unsigned int> nTrackCandidatespT5_buf;
-    Buf<TDev, unsigned int> nTrackCandidatespLS_buf;
-    Buf<TDev, unsigned int> nTrackCandidatesT5_buf;
-
-    Buf<TDev, uint8_t> logicalLayers_buf;
-    Buf<TDev, unsigned int> hitIndices_buf;
-    Buf<TDev, int> pixelSeedIndex_buf;
-    Buf<TDev, uint16_t> lowerModuleIndices_buf;
-
-    Buf<TDev, FPX> centerX_buf;
-    Buf<TDev, FPX> centerY_buf;
-    Buf<TDev, FPX> radius_buf;
-
-    TrackCandidates data_;
-
-    template <typename TQueue, typename TDevAcc>
-    TrackCandidatesBuffer(unsigned int maxTrackCandidates, TDevAcc const& devAccIn, TQueue& queue)
-        : trackCandidateType_buf(allocBufWrapper<short>(devAccIn, maxTrackCandidates, queue)),
-          directObjectIndices_buf(allocBufWrapper<unsigned int>(devAccIn, maxTrackCandidates, queue)),
-          objectIndices_buf(allocBufWrapper<unsigned int>(devAccIn, 2 * maxTrackCandidates, queue)),
-          nTrackCandidates_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
-          nTrackCandidatespT3_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
-          nTrackCandidatespT5_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
-          nTrackCandidatespLS_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
-          nTrackCandidatesT5_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
-          logicalLayers_buf(allocBufWrapper<uint8_t>(devAccIn, Params_pT5::kLayers * maxTrackCandidates, queue)),
-          hitIndices_buf(allocBufWrapper<unsigned int>(devAccIn, Params_pT5::kHits * maxTrackCandidates, queue)),
-          pixelSeedIndex_buf(allocBufWrapper<int>(devAccIn, maxTrackCandidates, queue)),
-          lowerModuleIndices_buf(allocBufWrapper<uint16_t>(devAccIn, Params_pT5::kLayers * maxTrackCandidates, queue)),
-          centerX_buf(allocBufWrapper<FPX>(devAccIn, maxTrackCandidates, queue)),
-          centerY_buf(allocBufWrapper<FPX>(devAccIn, maxTrackCandidates, queue)),
-          radius_buf(allocBufWrapper<FPX>(devAccIn, maxTrackCandidates, queue)) {
-      alpaka::memset(queue, nTrackCandidates_buf, 0u);
-      alpaka::memset(queue, nTrackCandidatesT5_buf, 0u);
-      alpaka::memset(queue, nTrackCandidatespT3_buf, 0u);
-      alpaka::memset(queue, nTrackCandidatespT5_buf, 0u);
-      alpaka::memset(queue, nTrackCandidatespLS_buf, 0u);
-      alpaka::memset(queue, logicalLayers_buf, 0u);
-      alpaka::memset(queue, lowerModuleIndices_buf, 0u);
-      alpaka::memset(queue, hitIndices_buf, 0u);
-      alpaka::memset(queue, pixelSeedIndex_buf, 0);
-    }
+  using TrackCandidatesDeviceCollection = PortableCollection<::lst::TrackCandidatesSoA>;
```

Comment:
Could you move the definition of `TrackCandidatesDeviceCollection` to a separate file, _e.g._ `RecoTracker/LSTCore/interface/alpaka/TrackCandidatesDeviceCollection.h` ?

Reply:
somewhat related, we were talking in an LST meeting to go the other way and move the `TrackCandidatesHostCollection.h` to be inside `TrackCandidatesSoA.h` .
Is it OK, or should they be kept separate?

Reply:
> somewhat related, we were talking in an LST meeting to go the other way and move the `TrackCandidatesHostCollection.h` to be inside `TrackCandidatesSoA.h` .
> Is it OK, or should they be kept separate?

One motivation to keep the `PortableCollection<T>` instantiations separate from the SoA definition would be to avoid compiling `PortableCollection` code with the device compilers, because the SoA definition alone should be sufficient for the needs of the device code.

Reply:
> One motivation to keep the `PortableCollection<T>` instantiations separate from the SoA definition would be to avoid compiling `PortableCollection` code with the device compilers, because the SoA definition alone should be sufficient for the needs of the device code.

This isn't necessarily very strong motivation, e.g. I'd guess we already have code that exposes `PortableCollection` code to the device compiler. But it is a consideration nevertheless.